prow/inrepoconfig/api/api.go

```diff
@@ -0,0 +1,151 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package api
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+
+	"github.com/sirupsen/logrus"
+	"sigs.k8s.io/yaml"
+
+	checkconfigapi "k8s.io/test-infra/prow/cmd/checkconfig/api"
+	"k8s.io/test-infra/prow/config"
+	"k8s.io/test-infra/prow/git"
+	"k8s.io/test-infra/prow/github"
+)
+
+const (
+	// PluginName is the name of this plugin
+	PluginName = "inrepoconfig"
+
+	// ContextName is the name of the context the inrepoconfig plugin creates at GitHub
+	ContextName = "prow-config-parsing"
+
+	// ConfigFileName is the name of the configfile the inrepoconfig plugin uses to read its
+	// config from
+	ConfigFileName = "prow.yaml"
+)
+
+// InRepoConfig is the type `prow.yaml` gets deserialized into
+type InRepoConfig struct {
+	Presubmits []config.Presubmit `json:"presubmits,omitempty"`
+}
+
+// defaultAndValidateInRepoConfig defaults and validates the InRepoConfig. This must be called before
+// accessing any data in it
+func (irc *InRepoConfig) defaultAndValidateInRepoConfig(c *config.Config, orgRepo string) error {
+	for i := range irc.Presubmits {
+		ps := &irc.Presubmits[i]
+		config.DefaultPresubmitFields(&c.ProwConfig, ps)
+		if ps.Decorate {
+			ps.DecorationConfig = ps.DecorationConfig.ApplyDefault(c.Plank.DefaultDecorationConfig)
+		}
+		if err := config.ResolvePresets(
+			ps.Name, ps.Labels, ps.Spec, ps.BuildSpec, c.Presets); err != nil {
+			return err
+		}
+	}
+
+	// Has to be done in the end, otherwise the jobs Trigger field is unset
+	if err := config.SetPresubmitRegexes(irc.Presubmits); err != nil {
+		return fmt.Errorf("failed to set Presubmit regexes: %v", err)
+	}
+
+	for _, presubmit := range irc.Presubmits {
+		if err := checkconfigapi.ValidatePresubmitJob(orgRepo, presubmit); err != nil {
+			return err
+		}
+	}
+	ap := &additionalPresubmits{
+		fullRepoName: orgRepo,
+		presubmits:   irc.Presubmits,
+	}
+	// This is not extremely efficient as it validates all Job types on all Repos of this
+	// Prow instance.
+	// Changing that requires a bigger refactoring of config.Config.ValidateJobConfig thought,
+	// in order to keep the change required when introducing `inrepoconfig` optimizing this is
+	// kept as a future task.
+	// TODO @alvaroaleman: Revisit and improve this
+	if err := c.ValidateJobConfig(ap); err != nil {
```

Comment:
Based on tracing that @cjwagner and I were doing, parsing out the regex in this step is actually some massive (50%+) part of our memory footprint so depending on how many PRs are being parsed in parallel for a repo as large as k/k this might actually matter.

Reply:
https://github.com/kubernetes/test-infra/pull/12836/files#r289911580